src/Plugin/Commerce/PriceRuleCalculation/AttributePrice.php

```diff
@@ -0,0 +1,227 @@
+      }
+      $form['price'][$key] = [
+        '#type' => 'details',
+        '#title' => $key,
+        '#collapsible' => TRUE,
+        '#open' => FALSE,
+      ];
+      $form['price'][$key]['field'] = [
+        '#type' => 'select',
+        '#title' => $this->t('Price Field'),
+        '#options' => $price_fields['commerce_price_fields'],
+        '#description' => $this->t('Select the field from which the attribute price needs to be fetched'),
+        '#default_value' => $field,
+      ];
+    }
+    return $form;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function submitConfigurationForm(array &$form, FormStateInterface $form_state) {
+    parent::submitConfigurationForm($form, $form_state);
+    $values = $form_state->getValue($form['#parents']);
+    $this->configuration['attribute_price'] = $values;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getLabel() {
+    return $this->t('Attribute Based Price');
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function calculate(
+    EntityInterface $entity,
+    PriceRuleInterface $price_rule,
+    $quantity,
+    Context $context
+  ) {
+    $this->assertEntity($entity);
+    return $this->recalculateVariationPrice($entity, $this->configuration['attribute_price']['price']);
+  }
+
+  /**
+   * Retrieves all attributes used by product variation types.
+   *
+   * @return array
+   *   Associative array which has product variation type as key and
+   *   the attributes used as value.
+   */
+  public function getAttributesPriceField() {
+    $price_fields = [];
+    $attribute_storage = $this->entityTypeManager->getStorage('commerce_product_attribute');
+    $attributes = $attribute_storage->loadMultiple();
+    foreach ($attributes as $attribute) {
+      $fields = \Drupal::service('entity_field.manager')->getFieldDefinitions('commerce_product_attribute_value', $attribute->id());
+      foreach ($fields as $field) {
+        $type = $field->getType();
+        if ($type == "commerce_price") {
+          $price_fields[$attribute->id()]['commerce_price_fields'][$field->get('field_name')] = $field->get('label');
+        }
+      }
+    }
+    return $price_fields;
+  }
+
+  /**
+   * Provides the total price of attribues attached to a variation.
+   *
+   * @param Drupal\commerce_product\Entity\ProductVariation $variation
+   *   The commerce product variation.
+   * @param array $attribute_field_map
+   *   The attribute price field mappings.
+   *
+   * @return Drupal\commerce_price\Price
+   *   Sum of all attribute values associated with a variation.
+   */
+  public function getAttributeTotalPrice(ProductVariation $variation, array $attribute_field_map) {
+    $attribute_total_price = new Price((string) 0, $variation->get('price')->first()->toPrice()->getCurrencyCode());
+    $attribute_storage = $this->entityTypeManager->getStorage('commerce_product_attribute');
+    $attributes = $attribute_storage->loadMultiple();
```

Comment:
**CHANGE REQUEST**

So here you will be loading all Attribute entities stored in the database, that can be a lot and it is unnecessary. There must be a better way, look if the `commerce_product` module has any helper functions, otherwise you can load the fields for the given variation type and get the ones that are entity reference fields configured to reference `ProductAttribute` entities.

Reply:
You are right. Modified to load only the attributes associated with the variation